datapath/link: Initialize link monitor explicitly

joestringer opened the pull request:
Previously, `pkg/datapath/link/` initialized an ad-hoc goroutine for
attempting to gather local link state information manually, and if an
error ever occurred, it would simply log the error and exit the
goroutine, rather than retrying.

Furthermore, this was all triggered from an `init()` function that would
directly initialize a package-level variable. This meant that there was
no way to mock out the implementation, and any code that imported the
link package would thereby create a direct dependency on the Linux
kernel (including any unit tests that use the package).

Factor out the implementation by introducing a constructor which creates
a controller to track the status of the link cache monitor, and by
explicitly initializing the link cache as part of the outermost binaries
that depend on it: cilium-agent and the in-pod cilium CLI. This makes
the dependencies more explicit, makes the logic easier to track at
runtime (eg via `cilium status`), more resilient to failures, allows the
dependencies to be tracked from the daemon more concretely, and reduces
platform dependencies in pkg/ code & tests.

joestringer commented on pkg/monitor/ifindex.go:
```diff
@@ -6,11 +6,23 @@ package monitor
+//
+// Package-level variable just to avoid overhauling old code to introduce an
+// overarching structure to help format the messages nicely.
+var linkMonitor getters.LinkGetter
```
I'm not the biggest fan of this, but at least it's abstracted so there's no direct dependency from pkg/monitor onto the Linux kernel. Unlike the `pkg/hubble` code, `pkg/monitor` doesn't seem to have the same notion of a parser structure that could hold the pointer to this (from the daemon), so I just manually plumbed a package-level variable here instead as baby steps towards cleaner code.

nathanjsweet commented on pkg/monitor/ifindex.go:
```diff
@@ -6,11 +6,23 @@ package monitor
+// overarching structure to help format the messages nicely.
+var linkMonitor getters.LinkGetter
+
+func Init(linkGetter getters.LinkGetter) {
```
nit: I suppose there's no way around having a public Init method, but can we at least leave a comment indicating that `cilium/cmd/monitor.go` is where it is called?

joestringer replied:
Yeah only real way around this would be to do more invasive refactoring of the `pkg/monitor` package to create an owning structure for parsing which could then be constructed using this object, then all of the in-package references to the linkMonitor would rely on the structure to fetch the pointer to the real copy of this object.

Personally I'm not a big fan of adding comments to places like this to say "This is called from X and Y", because those comments can easily get out of date when the other code is refactored. If someone is reading this code and really wants to know where it's called, then hopefully they have an editor that tells them this information, or they can combine the package name `monitor` with the function name `Init` to get a phrase they can grep for: 

```
$ git grep monitor.Init
cilium/cmd/monitor.go:  monitor.Init(linkCache)
```

However, I looked at this again and figured I'll just spend 20m refactoring monitor package so that we can just eliminate this package-level variable and corresponding initialization function. Will push a new version in a bit. Need to rebase anyhow.

joestringer commented on pkg/monitor/datapath_debug.go:
```diff
@@ -238,12 +239,12 @@ func (n *DebugMsg) DumpInfo(data []byte) {
 // Dump prints the debug message in a human readable format.
-func (n *DebugMsg) Dump(prefix string) {
-	fmt.Printf("%s MARK %#x FROM %d DEBUG: %s\n", prefix, n.Hash, n.Source, n.Message())
+func (n *DebugMsg) Dump(prefix string, linkMonitor getters.LinkGetter) {
```
^^ I briefly considered other ways to try to avoid passing this parameter down through these methods, like embedding it in the `DebugMsg`. My main concern was that I believe that `DebugMsg` is binary-compatible with the datapath BPF C structures that generate these notifications, so we can't just add another pointer somewhere here as this could potentially cause us to either inadvertently write to a field that we shouldn't (thereby messing up some message in the buffer), or plausibly interpet some random bytes as a pointer. This seemed like the simpler / safer option.

aanm replied:
The problem seems to be that `DebugMsg` is used as receiver for this method. We could use a structure that had 2 fields, a `DebugMsg` and a `linkMonitor`, then this new structure could have the `Dump` method instead.

joestringer replied:
yeah as I understand it, this is the role that the `MonitorFormatter` is trying to fill. Arguably we could shuffle the code around to attach this to the MonitorFormatter and hide these parameters from going down through a couple of layers of functions. At this point though I don't see any real benefit to going through all that.

aanm commented on pkg/datapath/link/link.go:
```diff
@@ -70,12 +55,29 @@ func GetIfIndex(ifName string) (uint32, error) {
-func (c *ifNameCache) syncCache() error {
+// NewLinkCache begins monitoring local interfaces for changes in order to
+// track local link information.
+func NewLinkCache() *LinkCache {
```
I don't know where this is called from but right now, the way `NewLinkCache` is written, this will behave as a singleton.

joestringer replied:
Ah, because the name for the controller is not unique? Yeah correct. Given that we will only call this constructor once from either the cilium-agent or `cilium monitor ...` commands, this is an OK tradeoff. I can make this more clear in the godoc.

joestringer replied:
^^ I ended up just adding a `sync.Once()` for this so we don't have to document. It means that there's still a package-level variable, but that's not such a big deal now that the LinkCache has a constructor and external packages need to create the LinkCache and use the methods against that object to fetch names.

aanm commented on pkg/datapath/link/link.go:
```diff
@@ -92,16 +94,20 @@ func (c *ifNameCache) syncCache() error {
+	if !ok {
+		name = fmt.Sprintf("%d", ifIndex)
+	}
+	return name
```
This looks like an innocent change but this might have performance implications in hubble which uses this function. I suggest to replace the `fmt.Sprintf` with a faster function. Probably `strconv.Itoa`.

joestringer replied:
The code is already doing this today, but while we're here we may as well fix this up yeah :+1: 

No idea how accurate this is but here's some random gist about performance of these alternatives:
https://gist.github.com/evalphobia/caee1602969a640a4530 

EDIT: Ah, the monitor path was using this but the hubble path wasn't. :+1:

aanm commented on pkg/hubble/parser/threefour/parser.go:
```diff
@@ -678,7 +680,7 @@ func decodeNetworkInterface(tn *monitor.TraceNotify, dbg *monitor.DebugCapture)
 	// if the interface is not found, `name` will be an empty string and thus
 	// omitted in the protobuf message
-	name, _ := link.GetIfNameCached(int(ifIndex))
+	name := p.linkGetter.GetIfNameCached(ifIndex)
```
This is the hubble function I'm referring to.

gandro commented on pkg/hubble/parser/threefour/parser.go:
```diff
@@ -678,7 +680,7 @@ func decodeNetworkInterface(tn *monitor.TraceNotify, dbg *monitor.DebugCapture)
 	// if the interface is not found, `name` will be an empty string and thus
 	// omitted in the protobuf message
-	name, _ := link.GetIfNameCached(int(ifIndex))
+	name := p.linkGetter.GetIfNameCached(ifIndex)
```
Nit: The old code was intentionally not converting `ifIndex` to a string if the interface wasn't found:

Hubble JSON output before this PR (in case ifIndex refers to a stale/unknown interface):
```javascript
  "interface": {
    "index": 25,
    // note: name omitted due to `omitempty`
  }
```

After this PR:
```javascript
  "interface": {
    "index": 25,
    "name": "25"
  }
```

I believe the `"name": "25"` introduced by this PR doesn't provide any value and therefore should be omitted.

joestringer replied:
Ah OK, thanks for the insight. I'll switch this up - I think I'll just revert the `GetIfNameCached` interface change here and maybe put the "always give me a string" version in a central location, maybe just part of the same interface so all the callers who want a string version can get it.

joestringer replied:
I debated just leaving the `ifname.go` file in the monitor package, but figured that once the name getter functions become methods as part of the LinkCache, it makes more sense to just move that code to the structure that it depends on. So in the end I added a new `Name()` method on the interface that allows getting this and always returning a string.

joestringer committed:
datapath/link: Add Name helper for link names

This commit shifts the ifName helper from the monitor package into the
central link package so it can be associated directly with the LinkCache
and allow other packages to get the interface in a human-friendly format
without having to hold dedicated translation code in their own packages.

Also converts the input into a uint32, which is the standard data type
for an ifindex on Linux, which will reduce the number of integer casts
necessary in outside code in order to interact with this function.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
datapath/link: Use controller to track link sync

Previously, pkg/datapath/link/ initialized an ad-hoc goroutine for
attempting to gather local link state information manually, and if an
error ever occurred, it would simply log the error and exit the
goroutine, rather than retrying.

Furthermore, this was all triggered from an init() function that would
directly initialize a package-level variable. This meant that there was
no way to mock out the implementation, and any code that imported the
link package would thereby create a direct dependency on the Linux
kernel (including any unit tests that use the package).

Factor out the implementation by introducing a constructor which creates
a controller to track the status of the link cache monitor. The public
constructor will be used to track the link cache in subsequent updates
and tidy up dependencies in other packages using dependency injection.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
datapath/link: Initialize link monitor explicitly

Explicitly initializing the link cache as part of the outermost binaries
that depend on it: cilium-agent and the in-pod cilium CLI. This makes
the dependencies more explicit, makes the logic easier to track at
runtime (eg via `cilium status`), more resilient to failures, allows the
dependencies to be tracked from the daemon more concretely, and reduces
platform dependencies in pkg/ code & tests.

The package-level variable introduced here in pkg/monitor to track the
LinkCache will be removed in a subsequent change, this is just helpful
for now to reduce the size of this particular patch.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
monitor: Remove package-level link monitor

Rather than forcing the cilium and cilium-agent binaries to initialize
the monitor package, just refactor the paths that make use of this
interface so that it can be passed down explicitly. This simplifies the
usage of the monitor package, removes a global variable, and reduces the
likelihood that this variable isn't properly initialized from any other
places.

Signed-off-by: Joe Stringer <joe@cilium.io>

gandro commented on pkg/hubble/testutils/fake.go:
```diff
@@ -326,6 +326,18 @@ var NoopEndpointGetter = FakeEndpointGetter{
+}
+
+func (e *FakeLinkGetter) GetIfNameCached(ifindex int) (string, bool) {
+	return e.Name(uint32(ifindex)), false
```
nit: Since this always returns `lo`

```suggestion
	return e.Name(uint32(ifindex)), true
```

gandro commented on pkg/datapath/link/link.go:
```diff
@@ -70,12 +65,34 @@ func GetIfIndex(ifName string) (uint32, error) {
+		linkCache = LinkCache{}
+
+		// Override the existing link-cache controller to update this copy
+		// we're currentyl constructing. If this function is called twice, then
```
```suggestion
		// we're currently constructing. If this function is called twice, then
```

joestringer replied:
:thumbsup: I don't think this comment was actually providing any value since I switched to using a sync.Once for initialization of the singleton, so I just dropped the whole comment.

gandro commented on pkg/hubble/parser/threefour/parser_test.go:
```diff
@@ -946,7 +947,7 @@ func TestTraceNotifyLocalEndpoint(t *testing.T) {
 	f := &flowpb.Flow{}
 
-	parser, err := New(log, &testutils.NoopEndpointGetter, &testutils.NoopIdentityGetter, &testutils.NoopDNSGetter, &testutils.NoopIPGetter, &testutils.NoopServiceGetter)
+	parser, err := New(log, &testutils.NoopEndpointGetter, &testutils.NoopIdentityGetter, &testutils.NoopDNSGetter, &testutils.NoopIPGetter, &testutils.NoopServiceGetter, &testutils.NoopLinkGetter)
```
nit:

Just a few lines below, we're obtaining the real `ifIndex` of the loopback device so we could test the old non-mockable getter. Now that we mock it, we could simplify it

The following could be replaced with `loIfIndex := 1`
https://github.com/cilium/cilium/blob/8e25d64386426ea97fa7b3a35f77d074ef4a7546/pkg/hubble/parser/threefour/parser_test.go#L956-L960

This if-statement could be removed:
https://github.com/cilium/cilium/blob/8e25d64386426ea97fa7b3a35f77d074ef4a7546/pkg/hubble/parser/threefour/parser_test.go#L991-L992

joestringer replied:
:+1: nice cleanup, I've added a new patch on top to address this.

gandro reviewed (commented):
Apologies for the late review. I found a few more nits (none of which need to be blocking however)

joestringer committed:
hubble/parser: Simplify unit test

This unit test previously relied on fetching the loopback device ifindex
on the node where it's run, but this is no longer necessary now that
link name fetching is abstracted by the previous commits.

Suggested-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Joe Stringer <joe@cilium.io>

gandro reviewed (approved):
Thanks a ton for the cleanup too!
